### PR TITLE
[4.0] Feature: Article > Save To Menu

### PR DESCRIPTION
**Pull request by HRIT-Florian**
Pull Request for Issue from facebook
https://www.facebook.com/groups/joomla.deutschland/permalink/1346440042196465/

### Summary of Changes
Save a content directy to a new menu item.

PR requires https://github.com/joomla/joomla-cms/pull/26675

Update: Removes also the old useless "add menu item" button.

### Testing Instructions
1. Apply Patch https://github.com/joomla/joomla-cms/pull/26675 (required)
2. Apply these Partch
3. Create a new article
4. Click on "Save to menu"
savetomenu
5. Article should been saved and redirects you to com_menu
6. Menu SingleArticle should been preselected with the article you've created and the title from the article

### Expected result
MenuNew

### Actual result
New Feature

### Documentation Changes Required
Yes, maybe

PR for the user survey from facebook + 50 users voted
SaveContent2Menu

### Additional Infos
**It's a global Save2Menu Function, could be finaly implement in any component**

### CODE REVIEW

**Comment by Fallatas** [comment]
I have tested this item :white_check_mark: successfully on 7a5d0c0036da625cf9a22d30d9c00f9f81c8b6ac<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26681">issues.joomla.org/tracker/joomla-cms/26681</a>.</sub>

**Comment by HRIT-Florian** [comment]
**It's a global Save2Menu Function, could be finaly implement in any component**

**Comment by 1apweb** [comment]
I have tested this item :red_circle: unsuccessfully on 7a5d0c0036da625cf9a22d30d9c00f9f81c8b6ac<br /><br />When I apply the patch there is this message : 
Warning: require(D:\xampp2\htdocs\pizza\libraries\vendor\composer/../hoa/consistency/Prelude.php): failed to open stream: No such file or directory in D:\xampp2\htdocs\pizza\libraries\vendor\composer\autoload_real.php on line 66

Fatal error: require(): Failed opening required 'D:\xampp2\htdocs\pizza\libraries\vendor\composer/../hoa/consistency/Prelude.php' (include_path='D:\xampp2\php\PEAR') in D:\xampp2\htdocs\pizza\libraries\vendor\composer\autoload_real.php on line 66<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26681">issues.joomla.org/tracker/joomla-cms/26681</a>.</sub>

File: administrator/language/en-GB/en-GB.ini [hunk 1]
```diff
@@ -958,6 +958,7 @@ JTOOLBAR_REFRESH_CACHE="Refresh Cache"
 JTOOLBAR_REMOVE="Remove"
 JTOOLBAR_SAVE="Save &amp; Close"
 JTOOLBAR_SAVE_AND_NEW="Save &amp; New"
+JTOOLBAR_SAVE_TO_MENU="Save to menu"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.ini]:
```suggestion
JTOOLBAR_SAVE_TO_MENU="Save to Menu"
```

**Comment by 1apweb** [comment]
When I apply the patch there is this message 

Warning: require(D:\xampp2\htdocs\pizza\libraries\vendor\composer/../hoa/consistency/Prelude.php): failed to open stream: No such file or directory in D:\xampp2\htdocs\pizza\libraries\vendor\composer\autoload_real.php on line 66

Fatal error: require(): Failed opening required 'D:\xampp2\htdocs\pizza\libraries\vendor\composer/../hoa/consistency/Prelude.php' (include_path='D:\xampp2\php\PEAR') in D:\xampp2\htdocs\pizza\libraries\vendor\composer\autoload_real.php on line 66<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26681">issues.joomla.org/tracker/joomla-cms/26681</a>.</sub>

File: administrator/components/com_content/Controller/ArticleController.php [hunk 2]
```diff
@@ -50,6 +51,45 @@ public function __construct($config = array(), MVCFactoryInterface $factory = nu
...
+	/**
+	 * Function that allows child controller access to model data
+	 * after the data has been saved.
+	 *
+	 * @param   BaseDatabaseModel  $model      The data model object.
+	 * @param   array              $validData  The validated data.
+	 *
+	 * @return  void
+	 *
+	 * @since   1.6
```
Comment (Quy) [inline administrator/components/com_content/Controller/ArticleController.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

**Comment by HRIT-Florian** [comment]
> 
> 
> When I apply the patch there is this message
> 
> Warning: require(D:\xampp2\htdocs\pizza\libraries\vendor\composer/../hoa/consistency/Prelude.php): failed to open stream: No such file or directory in D:\xampp2\htdocs\pizza\libraries\vendor\composer\autoload_real.php on line 66
> 
> Fatal error: require(): Failed opening required 'D:\xampp2\htdocs\pizza\libraries\vendor\composer/../hoa/consistency/Prelude.php' (include_path='D:\xampp2\php\PEAR') in D:\xampp2\htdocs\pizza\libraries\vendor\composer\autoload_real.php on line 66
> This comment was created with the [J!Tracker Application](https://github.com/joomla/jissues) at [issues.joomla.org/tracker/joomla-cms/26681](https://issues.joomla.org/tracker/joomla-cms/26681).

Your Error has nothing to do with the patch

**Comment by joomlamarco** [comment]
I have tested this item :white_check_mark: successfully on acf4bf1d6aa78c4011e51ffe8dbb35175802d9dc<br /><br />good feature<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26681">issues.joomla.org/tracker/joomla-cms/26681</a>.</sub>

**Comment by crommie** [comment]
I have tested this item :white_check_mark: successfully on 82859f47f6543d1461421c740e10ee61ae8baac2<br /><br />Yes! Nice.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26681">issues.joomla.org/tracker/joomla-cms/26681</a>.</sub>

**Comment by brianteeman** [comment]
This feature is already present
image

**Comment by HenkZ** [comment]
I have tested this item :red_circle: unsuccessfully on acf4bf1d6aa78c4011e51ffe8dbb35175802d9dc<br /><br />The following error message appaers by  installaing the requested pr #26675  ( ! ) Warning: require(C:\wamp\www\joomla\libraries\vendor\composer/../hoa/consistency/Prelude.php): failed to open stream: No such file or directory in C:\wamp\www\joomla\libraries\vendor\composer\autoload_real.php on line 66

I've installad Joomla_4.0.0-beta1-dev-Development-Full_Package nightly build and 
https://github.com/joomla-extensions/patchtester/releases/tag/4.0.0-beta2

So the same error as mentioned before.
Note: The autoload_real.php on line 66 seems an issue because other "pizza bug & fun attendees" have the save issue in Nijmegen (with other issues).



<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26681">issues.joomla.org/tracker/joomla-cms/26681</a>.</sub>

**Comment by HRIT-Florian** [comment]
@brianteeman only in edit an exisiting and not with passing data

**Comment by brianteeman** [comment]
Maybe but this is not the solution as we now have two completely different "save to menu" type buttons with completely different styling and different code

**Comment by sanderpotjer** [comment]
@Didldu-Florian I was guessing that as answer indeed. However, it might be a bit confusing now, as the behaviour is now different for the same action (creating menu item) depending on if your article is saved (modal) or not yet (redirect). Can't we make the behaviour similar to both cases?

I would also make more sense to add it to the "Save & Close" dropdown, so "Save & Create menu" or something?

**Comment by HRIT-Florian** [comment]
Yes, let me merge these in these PR, because the current feature is not helpful. It opens just the menu creatin in a modal and did not pass any data. Maybe we can use the button including my function. Will try to merge it.

**Comment by HRIT-Florian** [comment]
@brianteeman did also removed the old add menu button

**Comment by 1apweb** [comment]
I have tested this item :white_check_mark: successfully on 88cfd501fe5038b2d0bee8c4a04715f611d1b8ba<br /><br />It's fine<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26681">issues.joomla.org/tracker/joomla-cms/26681</a>.</sub>

**Comment by uglyeoin** [comment]
So we are waiting for new code to test again?

**Comment by anibalsanchez** [comment]
I have tested this item :white_check_mark: successfully on 88cfd501fe5038b2d0bee8c4a04715f611d1b8ba<br /><br />Test OK<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26681">issues.joomla.org/tracker/joomla-cms/26681</a>.</sub>

**Comment by HRIT-Florian** [comment]
@uglyeoin we can go finaly RTC
It's done for the core and com_content.

I will be working on implementing it to the other components, but not in these PR
So, can anybody please mark it as RTC? Thanks.

**Comment by SharkyKZ** [comment]
Sorry for late response. But this is not ready yet. Please remove this code, it was used by menu modal button https://github.com/joomla/joomla-cms/blob/823e9c9bfa837297c83ae65b9ecf4c1dff0a0348/administrator/components/com_content/View/Article/HtmlView.php#L211-L234

**Comment by brianteeman** [comment]
An observation - with the way workflows is setup now by default when you create an article it is in the unpublished state. So when you create a new article and save to menu the menu will 404 because the article it is linking to is not published yet

File: administrator/components/com_content/Controller/ArticleController.php [hunk 3]
```diff
@@ -50,6 +51,45 @@ public function __construct($config = array(), MVCFactoryInterface $factory = nu
...
+			$type = 'component';
+
+			$editState['id'] = $id;
+			$editState['link']  = $link;
+			$editState['title'] = $model->getItem($id)->title;
+			$editState['type']  = $type;
+			$editState['request']['id'] = $id;
+
+			$this->app->setUserState('com_menus.edit.item', array(
+				'id' => $id,
```
Comment (SharkyKZ) [inline administrator/components/com_content/Controller/ArticleController.php]: The ID here is of new menu item, not of article. I think `0` is fine.

Comment (HRIT-Florian) [reply administrator/components/com_content/Controller/ArticleController.php]: solved, thanks!

**Comment by HRIT-Florian** [comment]
Can we go RTC ?

**Comment by SharkyKZ** [comment]
This needs tests again after last code changes.

**Comment by brianteeman** [comment]
One thing I just noticed is that because the default behaviour of a new article is for it to be unpublished on save if you "save to menu" then it will appear to not work on the frontend because the article is not published

File: administrator/language/en-GB/en-GB.ini [hunk 4]
```diff
@@ -963,6 +963,7 @@ JTOOLBAR_REFRESH_CACHE="Refresh Cache"
 JTOOLBAR_REMOVE="Remove"
 JTOOLBAR_SAVE="Save &amp; Close"
 JTOOLBAR_SAVE_AND_NEW="Save &amp; New"
+JTOOLBAR_SAVE_TO_MENU="Save to Menu"
```
Comment (SharkyKZ) [inline administrator/language/en-GB/en-GB.ini]: Sort keys alphabetically.

**Comment by uglyeoin** [comment]
@Didldu-Florian I'll test when I have time, if you don't hear in a few days feel free to prompt me.

File: administrator/components/com_content/Controller/ArticleController.php [hunk 5]
```diff
@@ -50,6 +51,44 @@ public function __construct($config = array(), MVCFactoryInterface $factory = nu
...
+	 * after the data has been saved.
+	 *
+	 * @param   BaseDatabaseModel  $model      The data model object.
+	 * @param   array              $validData  The validated data.
+	 *
+	 * @return  void
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	protected function postSaveHook(BaseDatabaseModel $model, $validData = array())
```
Comment (Quy) [inline administrator/components/com_content/Controller/ArticleController.php:65]:
```suggestion
	protected function postSaveHook(BaseDatabaseModel $model, $validData = [])
```

File: administrator/components/com_content/Controller/ArticleController.php [hunk 6]
```diff
@@ -50,6 +51,44 @@ public function __construct($config = array(), MVCFactoryInterface $factory = nu
...
+			$link = 'index.php?option=com_content&view=article';
+			$type = 'component';
+
+			$editState['id'] = $id;
+			$editState['link']  = $link;
+			$editState['title'] = $model->getItem($id)->title;
+			$editState['type']  = $type;
+			$editState['request']['id'] = $id;
+
+			$this->app->setUserState('com_menus.edit.item', array(
```
Comment (Quy) [inline administrator/components/com_content/Controller/ArticleController.php:82]:
```suggestion
			$this->app->setUserState('com_menus.edit.item', [
```

File: administrator/components/com_content/Controller/ArticleController.php [hunk 7]
```diff
@@ -50,6 +51,44 @@ public function __construct($config = array(), MVCFactoryInterface $factory = nu
...
+				'data' => $editState,
+				'type' => $type,
+				'link' => $link)
```
Comment (Quy) [inline administrator/components/com_content/Controller/ArticleController.php:85]:
```suggestion
				'link' => $link
```

File: administrator/components/com_content/Controller/ArticleController.php [hunk 8]
```diff
@@ -50,6 +51,44 @@ public function __construct($config = array(), MVCFactoryInterface $factory = nu
...
+			);
```
Comment (Quy) [inline administrator/components/com_content/Controller/ArticleController.php:86]:
```suggestion
			]);
```

**Comment by vaibhavsTekdi** [comment]
I have tested this item :white_check_mark: successfully on 08dfd51f8f1496c8b1a7dbb5daf78807fdebd59a<br /><br />#jpbf19 #test1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26681">issues.joomla.org/tracker/joomla-cms/26681</a>.</sub>

**Comment by atpatil11** [comment]
I have tested this item :white_check_mark: successfully on 08dfd51f8f1496c8b1a7dbb5daf78807fdebd59a<br /><br />I have tested this Patch. It works!

This is simple & sweet feature.

#pbf19 #test2<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26681">issues.joomla.org/tracker/joomla-cms/26681</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26681">issues.joomla.org/tracker/joomla-cms/26681</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**